File: attached_assets/_extracted/Competency-Platform-main/src/server/services/scoring.ts
/**
 * Evaluation scoring engine.
 *
 * Ported from the prototype's `evRecalc`/`evTo100`/`evAvg` logic and made into a
 * pure, framework-free function so it can be unit-tested in isolation and reused
 * by the evaluation tRPC router (Phase 4).
 *
 * Scale rules (unchanged from the prototype):
 *  - Competencies are rated 1..5 and converted to a /100 score via (v-1)/4*100.
 *  - Shared competency groups: behavioral, leadership, technical. Plus optional
 *    job-specific competencies.
 *  - KPIs are entered directly as 0..100 achievement percentages.
 *  - competencyScore = mean of the present group /100 scores.
 *  - total = kpiScore*kpiWeight + competencyScore*(1-kpiWeight), rounded.
 *    If only one side is present, that side is used at full weight.
 */

// Mirrors prisma `EvaluationMode` without importing the generated client, keeping
// this module dependency-free and testable before `prisma generate` has run.
export type EvaluationMode = "SHARED" | "SPECIFIC" | "BOTH";

/** Stable refKeys for the shared competency groups, matching the prototype. */
export const SHARED_COMPETENCY_KEYS = {
  behavioral: ["b1", "b2", "b3", "b4", "b5"],
  leadership: ["l1", "l2", "l3", "l4", "l5"],
  technical: ["t1", "t2", "t3", "t4"],
} as const;

export interface ScoreInput {
  mode: EvaluationMode;
  /** KPI weight as a percentage 0..100. Competency weight is the remainder. */
  kpiWeight: number;
  /** Rated behavioral competency values (1..5). Only include rated ones. */
  behavioral?: number[];
  /** Rated leadership competency values (1..5). */
  leadership?: number[];
  /** Rated technical competency values (1..5). */
  technical?: number[];
  /** Rated job-specific competency values (1..5). */
  jobSpecific?: number[];
  /** KPI achievement percentages (0..100). */
  kpis?: number[];
}

export interface ScoreResult {
  kpiScore: number | null;
  competencyScore: number | null;
  groupScores: {
    behavioral: number | null;
    leadership: number | null;
    technical: number | null;
    jobSpecific: number | null;
  };
  totalScore: number | null;
  ratingLabel: string | null;
}

/** Rating bands (mirrors the prototype thresholds). */
export const RATING_BANDS = [
  { min: 91, label: "متميز" },
  { min: 76, label: "يتجاوز التوقعات" },
  { min: 61, label: "يحقق التوقعات" },
  { min: 41, label: "يحتاج تحسيناً" },
  { min: 0, label: "دون المستوى" },
] as const;

export function ratingLabelFor(score: number): string {
  for (const band of RATING_BANDS) {
    if (score >= band.min) return band.label;
  }
  return RATING_BANDS[RATING_BANDS.length - 1]!.label;
}

/** Convert a 1..5 competency rating to a /100 score. */
export function competencyTo100(value: number): number {
  return Math.round(((value - 1) / 4) * 100);
}

function mean(values: number[]): number | null {
  if (!values.length) return null;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

/** Average a group of 1..5 ratings and convert to /100, or null if empty. */
function groupTo100(values: number[] | undefined): number | null {
  const avg = mean(values ?? []);
  return avg === null ? null : competencyTo100(avg);
}

export function calculateScore(input: ScoreInput): ScoreResult {
  const useShared = input.mode === "SHARED" || input.mode === "BOTH";
  const useSpecific = input.mode === "SPECIFIC" || input.mode === "BOTH";

  const behavioral = useShared ? groupTo100(input.behavioral) : null;
  const leadership = useShared ? groupTo100(input.leadership) : null;
  const technical = useShared ? groupTo100(input.technical) : null;
  const jobSpecific = useSpecific ? groupTo100(input.jobSpecific) : null;

  const presentGroups = [behavioral, leadership, technical, jobSpecific].filter(
    (v): v is number => v !== null,
  );
  const competencyScore = presentGroups.length
    ? presentGroups.reduce((a, b) => a + b, 0) / presentGroups.length
    : null;

  const kpiScore = mean(input.kpis ?? []);

  const kpiWeight = clampWeight(input.kpiWeight) / 100;
  const competencyWeight = 1 - kpiWeight;

  let total: number | null = null;
  if (kpiScore !== null && competencyScore !== null) {
    total = kpiScore * kpiWeight + competencyScore * competencyWeight;
  } else if (kpiScore !== null) {
    total = kpiScore;
  } else if (competencyScore !== null) {
    total = competencyScore;
  }

  const totalScore = total === null ? null : Math.round(total);

  return {
    kpiScore,
    competencyScore,
    groupScores: { behavioral, leadership, technical, jobSpecific },
    totalScore,
    ratingLabel: totalScore === null ? null : ratingLabelFor(totalScore),
  };
}

function clampWeight(weight: number): number {
  if (Number.isNaN(weight)) return 60;
  return Math.min(100, Math.max(0, weight));
}

File: artifacts/api-server/internal/scoring/scoring.go
// Package scoring ports the evaluation scoring engine from the original
// TypeScript service. Pure and dependency-free so it can be unit tested.
package scoring

import "math"

// Mode mirrors the EvaluationMode enum.
type Mode string

const (
	ModeShared   Mode = "SHARED"
	ModeSpecific Mode = "SPECIFIC"
	ModeBoth     Mode = "BOTH"
)

// SharedCompetencyKeys are the stable refKeys for the shared competency groups.
var SharedCompetencyKeys = map[string][]string{
	"behavioral": {"b1", "b2", "b3", "b4", "b5"},
	"leadership": {"l1", "l2", "l3", "l4", "l5"},
	"technical":  {"t1", "t2", "t3", "t4"},
}

// RatingBand maps a minimum score to an Arabic rating label.
type RatingBand struct {
	Min   int
	Label string
}

var RatingBands = []RatingBand{
	{91, "متميز"},
	{76, "يتجاوز التوقعات"},
	{61, "يحقق التوقعات"},
	{41, "يحتاج تحسيناً"},
	{0, "دون المستوى"},
}

// RatingLabelFor returns the Arabic rating band label for a score.
func RatingLabelFor(score int) string {
	for _, b := range RatingBands {
		if score >= b.Min {
			return b.Label
		}
	}
	return RatingBands[len(RatingBands)-1].Label
}

// CompetencyTo100 converts a 1..5 rating to a /100 score.
func CompetencyTo100(value float64) int {
	return int(math.Round(((value - 1) / 4) * 100))
}

func mean(values []float64) (float64, bool) {
	if len(values) == 0 {
		return 0, false
	}
	var sum float64
	for _, v := range values {
		sum += v
	}
	return sum / float64(len(values)), true
}

func groupTo100(values []float64) *int {
	avg, ok := mean(values)
	if !ok {
		return nil
	}
	v := CompetencyTo100(avg)
	return &v
}

// Input describes the rated values for a single evaluation.
type Input struct {
	Mode        Mode
	KpiWeight   float64 // 0..100
	Behavioral  []float64
	Leadership  []float64
	Technical   []float64
	JobSpecific []float64
	Kpis        []float64 // 0..100 achievement percentages
}

// GroupScores holds the per-group /100 scores (nil when absent).
type GroupScores struct {
	Behavioral  *int `json:"behavioral"`
	Leadership  *int `json:"leadership"`
	Technical   *int `json:"technical"`
	JobSpecific *int `json:"jobSpecific"`
}

// Result is the computed scoring outcome.
type Result struct {
	KpiScore        *float64    `json:"kpiScore"`
	CompetencyScore *float64    `json:"competencyScore"`
	GroupScores     GroupScores `json:"groupScores"`
	TotalScore      *int        `json:"totalScore"`
	RatingLabel     *string     `json:"ratingLabel"`
}

func clampWeight(w float64) float64 {
	if math.IsNaN(w) {
		return 60
	}
	return math.Min(100, math.Max(0, w))
}

// Calculate computes the full scoring result, matching the original logic:
//   - competency groups: mean of 1..5 ratings -> /100
//   - competencyScore: mean of present groups
//   - kpiScore: mean of kpi achievements
//   - total = kpiScore*w + competencyScore*(1-w), one side -> full weight.
func Calculate(in Input) Result {
	useShared := in.Mode == ModeShared || in.Mode == ModeBoth
	useSpecific := in.Mode == ModeSpecific || in.Mode == ModeBoth

	var behavioral, leadership, technical, jobSpecific *int
	if useShared {
		behavioral = groupTo100(in.Behavioral)
		leadership = groupTo100(in.Leadership)
		technical = groupTo100(in.Technical)
	}
	if useSpecific {
		jobSpecific = groupTo100(in.JobSpecific)
	}

	var present []float64
	for _, g := range []*int{behavioral, leadership, technical, jobSpecific} {
		if g != nil {
			present = append(present, float64(*g))
		}
	}

	var competencyScore *float64
	if len(present) > 0 {
		cs, _ := mean(present)
		competencyScore = &cs
	}

	var kpiScore *float64
	if ks, ok := mean(in.Kpis); ok {
		kpiScore = &ks
	}

	w := clampWeight(in.KpiWeight) / 100
	cw := 1 - w

	var total *float64
	switch {
	case kpiScore != nil && competencyScore != nil:
		t := *kpiScore*w + *competencyScore*cw
		total = &t
	case kpiScore != nil:
		total = kpiScore
	case competencyScore != nil:
		total = competencyScore
	}

	var totalScore *int
	var ratingLabel *string
	if total != nil {
		ts := int(math.Round(*total))
		totalScore = &ts
		rl := RatingLabelFor(ts)
		ratingLabel = &rl
	}

	return Result{
		KpiScore:        kpiScore,
		CompetencyScore: competencyScore,
		GroupScores:     GroupScores{behavioral, leadership, technical, jobSpecific},
		TotalScore:      totalScore,
		RatingLabel:     ratingLabel,
	}
}
